### PR TITLE
Make scheduler configurable on ReactiveAuthenticationManagerAdapter

### PR DESCRIPTION
**Pull request by ttddyy**
Currently, authentication logic will be performed on hardcoded elastic scheduler in `ReactiveAuthenticationManagerAdapter`.
This needs to be configurable in order to perform the authentication logic in dedicated scheduler.

Also, once authentication is performed, rest of the logic(Controller/Service) also be performed on the published scheduler.
To allow the rest of the logic after authentication to run on different scheduler, adding another optional scheduler to publish on.

This PR is based on the usecase I have described here: https://github.com/reactor/reactor-core/pull/1644#issuecomment-490701518

### CODE REVIEW

**Comment by rwinch** [comment]
Thanks for the PR @ttddyy!

Can you please help me better understand what problems you are trying to solve?

**Comment by ttddyy** [comment]
Hi @rwinch 

The change has two aspect.

First one is that we want to run the auth logic on controlled executor rather than the elastic one. Currently, it is hardcoded, so I want to make it changeable.

The second point is after auth process has finished, we would like to switch the executor. Currently, since auth logic has switched to the elastic executor, subsequent logic (such as performing controller/service) also continue to be executed on the switched elastic executor.
In my case, after performing auth on the controlled executor, now we want to switch to another executor that is managed differently to perform application logic.
(Switching back may be debatable action to do, according to the discussion on https://github.com/reactor/reactor-core/pull/1644, but having such capability is useful IMO)

**Comment by rwinch** [comment]
Ok I understand the what, but why are you needing to do this? Specifically, why do you need to switch back? What are you switching back to?

**Comment by ttddyy** [comment]
I would like to publish to another executor that executes application logic. The main reason is to have metrics around the executors that perform auth logic and app logic *separately*. This way we can get independent metrics for auth and make decision on converting the auth part to full reactive or just use adapter approach.

**Comment by rwinch** [comment]
Can't you specify the executor for the logic you are executing? Why does this need to be part of Spring Security's code?

**Comment by ttddyy** [comment]
OK, yeah executor for application can be specified outside of spring-security such as another webfilter or so. I'll update the PR to only focus on making auth-manager scheduler modifiable.

File: core/src/main/java/org/springframework/security/authentication/ReactiveAuthenticationManagerAdapter.java [hunk 1]
```diff
@@ -50,4 +54,13 @@ public ReactiveAuthenticationManagerAdapter(AuthenticationManager authentication
 			})
 			.filter( a -> a.isAuthenticated());
 	}
+
+	/**
+	 * Set a scheduler that will be published on to perform the authentication logic.
+	 * @param scheduler a scheduler to be published on
+	 */
+	public void setScheduler(Scheduler scheduler) {
```
Comment (rwinch) [inline core/src/main/java/org/springframework/security/authentication/ReactiveAuthenticationManagerAdapter.java]: Please add a null check and a test to validate the null check

Comment (ttddyy) [reply core/src/main/java/org/springframework/security/authentication/ReactiveAuthenticationManagerAdapter.java]: Updated the PR with null check and test

**Comment by ttddyy** [comment]
Hi @rwinch
This PR is ready, can you care to merge?
Thanks,

**Comment by rwinch** [comment]
Thanks for the PR and the ping @ttddyy! This is now merged into master
